teps/0158-pipeline-fail-fast.md

```diff
@@ -0,0 +1,160 @@
+- [Motivation](#motivation)
+  * [Goals](#goals)
+  * [use-cases](#use-cases)
+- [Proposal](#proposal)
+    * [failFast](#failfast)
+    * [config-defaults](#config-defaults)
+    * [PipelineRun](#pipelinerun)
+    * [Pipeline](#pipeline)
+- [Design Evaluation](#design-evaluation)
+    * [Priority](#priority)
+    * [Performance](#performance)
+- [References](#references)
+<!-- /toc -->
+
+
+## Summary
+
+This proposal is to support stopping the execution of the Pipeline immediately when a Task in the Pipeline failed.
+Because there may be multiple parallel tasks in the pipeline, when one task fails,
+The final status of the pipeline is Failed, but other tasks may still be executed,
+which will waste resources. Therefore, we need a mechanism to support the
+immediate stop of pipeline execution when a task failed.
+
+## Motivation
+
+When a Task in a Pipeline fails, the Pipeline status may eventually change to Failed, but other parallel Tasks continue to execute. 
+When the final Pipeline status is Failed, There is a high probability that users will create a new Pipeline execution,
+because in many cases they hope that the final Pipeline status will be Success. Therefore, 
+in this case, it is necessary to stop the Pipeline execution as soon as possible. The fail-fast mechanism can 
+reduce the waste of resources on the one hand, and on the other hand, 
+it can also let users know the execution results of the Pipeline as soon as possible.
+
+### Goals
+
+- Support canceling the execution of the Pipeline immediately when a Task in the Pipeline failed.
+
+### use-cases
+
+Take the following pipeline as an example. When `fail-task` fails to execute, 
+the parallel `success1` and `success2`Tasks will exit execution.
+The status of PipelineRun is `Failed`, the status of `fail-task` is `Failed`, 
+and the status of `success1` and `success2` are `RunCancelled`.
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: PipelineRun
+metadata:
+  name: pipeline-run
+spec:
+  failFast: "true"
+  pipelineSpec:
+    tasks:
+    - name: fail-task
+      taskSpec:
+        steps:
+          - name: fail-task
+            image: busybox
+            command: ["/bin/sh", "-c"]
+            args:
+              - exit 1
+    - name: success1
+      taskSpec:
+        steps:
+          - name: success1
+            image: busybox
+            command: ["/bin/sh", "-c"]
+            args:
+              - sleep 360
+    - name: success2
+      taskSpec:
+        steps:
+          - name: success2
+            image: busybox
+            command: ["/bin/sh", "-c"]
+            args:
+              - sleep 360
+```
+
+## Proposal
+
+### failFast
+
+The `failFast` property is of string type and supports the following values:
+1. `true`: When a Task in the Pipeline fails, the Pipeline execution is stopped immediately.
+2. `false`: This is consistent with the current behavior. When a Task in a Pipeline fails, the Pipeline status changes to Failed, but other Tasks continue to execute.
```

Comment:
This sounds like we have fail-fast half implemented... I think the two options ideally should be:
- `true`: When something causes the pipeline run to fail, all running tasks are terminated immediately 
- `false`: When something causes the pipeline run to fail, no new tasks are scheduled. Once all running tasks are finished, the pipeline is marked as failed

But I guess we could keep `false` to behave like it does today and perhaps change it as part of a separate effort.
It would be nice for this to have graceful termination of Tasks and we don't want to lose logs in case of fail fast

Reply:
> It would be nice for this to have graceful termination of Tasks and we don't want to lose logs in case of fail fast

graceful termination has been implemented by [feat/Cancel taskrun using entrypoint binary](https://github.com/tektoncd/pipeline/pull/6511)

and it will be promoted to default